### PR TITLE
issue #143 solved✨

### LINKED ISSUE #143: [Feature Request] <adding Combinatorics>
### Is your feature request related to a problem? Please describe.
this is one more important part of DSA ,ive seen loads of questions on this in leetcode aswell. I would like to add this as well.

**Would you like to work on this feature?**  
- [X] Yes  
- [ ] No

### PR DESCRIPTION
**Pull request by trahulprabhu38**
### Description 
issue #143 added combinatorics.

@ajay-dhangar please review this PR.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Documentation update

### Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: docs/combinatorics/combinatorics-theory.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
+---
+id: combinatorics
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: The id field in this file must be changed. Each route requires a unique id, and the current ones are not unique.

File: docs/combinatorics/combinatorics-theory.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+## Introduction to Combinatorics
+
+Combinatorics is a branch of mathematics dealing with combinations, arrangements, and counting of objects. In the context of Data Structures and Algorithms (DSA), combinatorial techniques are used to solve problems related to counting, arrangement, and selection of data. These techniques are crucial in algorithm design and analysis, particularly in fields like graph theory, probability, and optimization.
+
+## Fundamental Concepts
+
+### 1. **Permutations**
+
+Permutations refer to the different arrangements of a set of elements. The number of permutations of \( n \) distinct objects is given by:
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: replace `\( n \)` with  `$n$`

File: docs/combinatorics/combinatorics-theory.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+\[
+P(n) = n!
+\]
+
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: replace

```md
[
P(n) = n!
\]
```

with
```md
$$
P(n) = n!
$$
```

File: docs/combinatorics/combinatorics-theory.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+where \( n! \) (n factorial) is the product of all positive integers up to \( n \).
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: replace `\( n! \)` with `$n!$` and `\( n \)` with `$n$`

File: docs/combinatorics/combinatorics-theory.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+**Example:**
+- The permutations of the set \(\{A, B, C\}\) are:
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: replace `\(\{A, B, C\}\)` with `$(A, B, C)$`

File: docs/combinatorics/combinatorics-theory.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+  - CBA
+
+### 2. **Combinations**
+
+Combinations refer to the selection of items from a larger set where the order does not matter. The number of ways to choose \( r \) elements from \( n \) elements is given by the binomial coefficient:
+
+\[
+C(n, r) = \frac{n!}{r!(n-r)!}
+\]
+
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: replace 
```md
\[
C(n, r) = \frac{n!}{r!(n-r)!}
\]
```
with

```md
$$
C(n, r) = \frac{n!}{r!(n-r)!}
$$
```

File: docs/combinatorics/combinatorics-theory.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+  - ABC
+  - ACB
+  - BAC
+  - BCA
+  - CAB
+  - CBA
+
+### 2. **Combinations**
+
+Combinations refer to the selection of items from a larger set where the order does not matter. The number of ways to choose \( r \) elements from \( n \) elements is given by the binomial coefficient:
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: replace `\( r \)` with `$r$` and `\( n \)` with `$n$`

File: docs/combinatorics/combinatorics-theory.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+\[
+C(n, r) = \frac{n!}{r!(n-r)!}
+\]
+
+**Example:**
+- The combinations of choosing 2 elements from the set \(\{A, B, C\}\) are:
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: replace `\(\{A, B, C\}\)` with `$(A, B, C)$`

File: docs/combinatorics/combinatorics-theory.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+  - AB
+  - AC
+  - BC
+
+### 3. **Binomial Theorem**
+
+The Binomial Theorem provides a formula for expanding expressions of the form \((x + y)^n\):
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: replace `\((x + y)^n\)` with `$(x + y)^n$`

File: docs/combinatorics/combinatorics-theory.md [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+\[
+(x + y)^n = \sum_{k=0}^{n} C(n, k) x^{n-k} y^k
+\]
+
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: replace 

```md
\[
(x + y)^n = \sum_{k=0}^{n} C(n, k) x^{n-k} y^k
\]
```

with

```md
$$
(x + y)^n = \sum_{k=0}^{n} C(n, k) x^{n-k} y^k
$$
```

File: docs/combinatorics/combinatorics-theory.md [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+Combinatorial counting is crucial in backtracking algorithms, especially in problems like:
+
+- **N-Queens Problem**: Counting valid arrangements of queens on a chessboard.
+- **Subset Sum Problem**: Finding combinations of numbers that sum to a given target.
+
+### 4. **Dynamic Programming**
+
+Dynamic programming often employs combinatorial principles to solve optimization problems by breaking them down into simpler subproblems. Examples include:
+
+- **Fibonacci Sequence**: Counting the number of ways to reach the \( n^{th} \) step.
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: replace `\( n^{th} \)` with `$ n^{th} $`

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
update your code based on suggestions

**Comment by trahulprabhu38** [comment]
resolved all problems.

**Comment by trahulprabhu38** [comment]
@ajay-dhangar all the required changes done.

File: docs/combinatorics/combinatorics-theory.md [hunk 12]
```diff
@@ -0,0 +1,112 @@
+---
+id: combinatorics-in-DataStructures
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-theory.md]: replace `id: combinatorics-in-DataStructures` with `id: combinatorics-theory`

File: docs/combinatorics/combinatorics-examples.md [hunk 13]
```diff
@@ -0,0 +1,123 @@
+---
+id: combinatorics-in-DataStructures
```
Comment (ajay-dhangar) [inline docs/combinatorics/combinatorics-examples.md]: replace `id: combinatorics-in-DataStructures` with `id: combinatorics-example`

**Review by ajay-dhangar — CHANGES_REQUESTED** [review]
update `id` based on suggestions to solve warning duplicate route found.

**Review by ajay-dhangar — APPROVED** [review]
Done

**Pull request merged**